src/System.Net.Primitives/src/System/Net/IPAddressParser.cs

```diff
@@ -54,24 +79,79 @@ internal static unsafe string IPv4AddressToString(uint address)
             char* addressString = stackalloc char[MaxLength];
-            int offset = MaxLength;
 
-            FormatIPv4AddressNumber((int)((address >> 24) & 0xFF), addressString, ref offset);
-            addressString[--offset] = '.';
-            FormatIPv4AddressNumber((int)((address >> 16) & 0xFF), addressString, ref offset);
-            addressString[--offset] = '.';
-            FormatIPv4AddressNumber((int)((address >> 8) & 0xFF), addressString, ref offset);
-            addressString[--offset] = '.';
+            int charsWritten = IPv4AddressToStringHelper(address, addressString);
+
+            return new string(addressString, 0, charsWritten);
+        }
+
+        internal static unsafe bool IPv4AddressToString(uint address, Span<char> formatted, out int charsWritten)
+        {
+            const int MaxLength = 15;
+
+            if (formatted.Length < MaxLength)
+            {
+                charsWritten = 0;
+                return false;
+            }
+     
+            fixed (char* addressString = &formatted.DangerousGetPinnableReference())
+            {
+                charsWritten = IPv4AddressToStringHelper(address, addressString);
+            }
+
+            return true;
+        }
+
+        internal static unsafe int IPv4AddressToStringHelper(uint address, char* addressString)
+        {
+            int offset = 0;
+
             FormatIPv4AddressNumber((int)(address & 0xFF), addressString, ref offset);
+            addressString[offset++] = '.';
+            FormatIPv4AddressNumber((int)((address >> 8) & 0xFF), addressString, ref offset);
+            addressString[offset++] = '.';
+            FormatIPv4AddressNumber((int)((address >> 16) & 0xFF), addressString, ref offset);
+            addressString[offset++] = '.';
+            FormatIPv4AddressNumber((int)((address >> 24) & 0xFF), addressString, ref offset);
 
-            return new string(addressString, offset, MaxLength - offset);
+            return offset;
         }
 
         internal static string IPv6AddressToString(ushort[] address, uint scopeId)
         {
             Debug.Assert(address != null);
             Debug.Assert(address.Length == IPAddressParserStatics.IPv6AddressShorts);
 
+            var buffer = Ipv6AddressToStringHelper(address, scopeId);
+
+            return StringBuilderCache.GetStringAndRelease(buffer);
+        }
+
+        internal static bool IPv6AddressToString(ushort[] address, uint scopeId, Span<char> destination, out int charsWritten)
+        {
+            Debug.Assert(address != null);
+            Debug.Assert(address.Length == IPAddressParserStatics.IPv6AddressShorts);
+
+            var buffer = Ipv6AddressToStringHelper(address, scopeId);
+
+            if (destination.Length < buffer.Length)
+            {
+                StringBuilderCache.Release(buffer);
+                charsWritten = 0;
+                return false;
+            }
+
+            buffer.CopyTo(0, destination, buffer.Length);
```

Comment:
It's nice to see how all of these span-based methods start layering on top of each other :smile: